File: vciplreportsv1001/vcipl1001/report/sales_order_fill_ratio/sales_order_fill_ratio.py
import copy
import frappe
from frappe import _
from frappe.utils import flt, date_diff


def execute(filters=None):

    if not filters:
        return [], [], None, []

    validate_filters(filters)

    columns = get_columns(filters)
    conditions = get_conditions(filters)
    data = get_data(conditions, filters)

    if not data:
        return columns, [], None, []

    data, chart_data, totals = prepare_data(data, filters)

    report_summary = [
        {"label": _("Ordered Qty"), "value": totals["qty"], "datatype": "Float"},
        {"label": _("Delivered Qty"), "value": totals["delivered_qty"], "datatype": "Float"},
        {"label": _("Pending Qty"), "value": totals["pending_qty"], "datatype": "Float"},
    ]

    return columns, data, None, chart_data, report_summary


# ---------------- VALIDATION ----------------

def validate_filters(filters):
    if filters.get("from_date") and filters.get("to_date"):
        if date_diff(filters.to_date, filters.from_date) < 0:
            frappe.throw(_("To Date cannot be before From Date."))


# ---------------- CONDITIONS ----------------

def get_conditions(filters):

    conditions = ""

    if filters.get("from_date") and filters.get("to_date"):
        conditions += " AND so.transaction_date BETWEEN %(from_date)s AND %(to_date)s"

    if filters.get("company"):
        conditions += " AND so.company = %(company)s"

    if filters.get("sales_order"):
        conditions += " AND so.name IN %(sales_order)s"

    if filters.get("warehouse"):
        conditions += " AND soi.warehouse = %(warehouse)s"

    return conditions


# ---------------- DATA ----------------

def get_data(conditions, filters):

    data = frappe.db.sql(f"""
        SELECT
            so.transaction_date AS date,
            so.name AS sales_order,
            so.status,
            c.customer_name AS customer,
            soi.item_code,
            soi.item_name,
            soi.qty,
            soi.delivered_qty,
            (soi.qty - soi.delivered_qty) AS pending_qty,
            soi.delivery_date,
            soi.warehouse
        FROM `tabSales Order` so
        INNER JOIN `tabSales Order Item` soi ON soi.parent = so.name
        INNER JOIN `tabCustomer` c ON c.name = so.customer
        WHERE so.docstatus = 1
        AND so.status NOT IN ('Completed','Closed','Stopped','On Hold')
        {conditions}
    """, filters, as_dict=True)

    bins = frappe.db.sql("""
        SELECT item_code, warehouse, actual_qty
        FROM `tabBin`
    """, as_dict=True)

    bin_map = {(b.item_code, b.warehouse): b.actual_qty for b in bins}

    so_popup = {}

    for row in data:

        available = bin_map.get((row.item_code, row.warehouse), 0)

        ratio = (row.delivered_qty / row.qty * 100) if row.qty > 0 else 0
        row["fill_ratio"] = round(ratio, 2)

        so_popup.setdefault(row.sales_order, {
            "items": [],
            "totals": {"ordered": 0, "delivered": 0, "pending": 0}
        })

        so_popup[row.sales_order]["items"].append({
            "item_name": row.item_name,
            "ordered_qty": row.qty,
            "delivered_qty": row.delivered_qty,
            "pending_qty": row.pending_qty,
            "available_qty": available,
            "ratio": round(ratio, 2),
            "delivery_date": str(row.delivery_date) if row.delivery_date else ""
        })

        so_popup[row.sales_order]["totals"]["ordered"] += flt(row.qty)
        so_popup[row.sales_order]["totals"]["delivered"] += flt(row.delivered_qty)
        so_popup[row.sales_order]["totals"]["pending"] += flt(row.pending_qty)

    for row in data:
        row["pending_popup"] = frappe.as_json(so_popup.get(row.sales_order, {}))

    return data


# ---------------- PREPARE ----------------

def prepare_data(data, filters):

    sales_order_map = {}

    totals = {"qty": 0, "delivered_qty": 0, "pending_qty": 0}

    for row in data:

        totals["qty"] += flt(row.qty)
        totals["delivered_qty"] += flt(row.delivered_qty)
        totals["pending_qty"] += flt(row.pending_qty)

        if filters.get("group_by_so"):

            so = row.sales_order

            if so not in sales_order_map:
                sales_order_map[so] = copy.deepcopy(row)
            else:
                sales_order_map[so].qty += flt(row.qty)
                sales_order_map[so].delivered_qty += flt(row.delivered_qty)
                sales_order_map[so].pending_qty += flt(row.pending_qty)

    for so in sales_order_map.values():
        so.fill_ratio = round(
            (so.delivered_qty / so.qty * 100) if so.qty > 0 else 0, 2
        )

    chart_data = {
        "data": {
            "labels": [_("Pending Qty"), _("Delivered Qty")],
            "datasets": [{"values": [totals["pending_qty"], totals["delivered_qty"]]}],
        },
        "type": "donut",
        "height": 300,
    }

    if filters.get("group_by_so"):
        return list(sales_order_map.values()), chart_data, totals

    return data, chart_data, totals


# ---------------- COLUMNS ----------------

def get_columns(filters):

    return [
        {"label": _("Date"), "fieldname": "date", "fieldtype": "Date"},
        {"label": _("Sales Order"), "fieldname": "sales_order", "fieldtype": "Link", "options": "Sales Order"},
        {"label": _("Pending Delivery"), "fieldname": "pending_delivery"},
        {"label": _("Customer"), "fieldname": "customer"},

        {"label": _("Ordered Qty"), "fieldname": "qty", "fieldtype": "Float"},
        {"label": _("Delivered Qty"), "fieldname": "delivered_qty", "fieldtype": "Float"},
        {"label": _("Pending Qty"), "fieldname": "pending_qty", "fieldtype": "Float"},
        {"label": _("Fill Ratio %"), "fieldname": "fill_ratio", "fieldtype": "Float"},

        {"fieldname": "pending_popup", "hidden": 1},
    ]

File: vciplreportsv1001/vcipl1001/report/sales_order_fill_ratio/sales_order_fill_ratio.js
// frappe.query_reports["SALES ORDER FILL RATIO"] = {

//     filters: [
//         { fieldname: "company", label: __("Company"), fieldtype: "Link", options: "Company", default: frappe.defaults.get_user_default("Company"), reqd: 1 },
//         { fieldname: "from_date", label: __("From Date"), fieldtype: "Date" },
//         { fieldname: "to_date", label: __("To Date"), fieldtype: "Date" },
//         { fieldname: "warehouse", label: __("Warehouse"), fieldtype: "Link", options: "Warehouse" },
//         { fieldname: "group_by_so", label: __("Group by Sales Order"), fieldtype: "Check", default: 1 }
//     ],

//     formatter(value, row, column, data, default_formatter) {

//         value = default_formatter(value, row, column, data);

//         if (column.fieldname === "pending_delivery") {
//             return `<a style="font-weight:bold;color:#1674E0;cursor:pointer"
//                 onclick='frappe.query_reports["SALES ORDER FILL RATIO"]
//                 .show_popup(${data.pending_popup}, "${data.customer}", "${data.sales_order}")'>
//                 View Items
//             </a>`;
//         }

//         return value;
//     },

//     show_popup(data, customer, sales_order) {

//         let rows = data.items || [];
//         let totals = data.totals || {};

//         let html = `
//         <div id="pending-popup">

//         <h3>Sales Order Fill Ratio</h3>
//         <h4>Customer: ${customer}</h4>
//         <h5>Sales Order: ${sales_order}</h5>

//         <table class="table table-bordered">
//         <tr>
//             <th>Item</th>
//             <th>Ordered</th>
//             <th>Delivered</th>
//             <th>Pending</th>
//             <th>Available</th>
//             <th>Fill %</th>
//             <th>Date</th>
//         </tr>`;

//         rows.forEach(r => {
//             html += `
//             <tr>
//                 <td>${r.item_name}</td>
//                 <td>${r.ordered_qty}</td>
//                 <td>${r.delivered_qty}</td>
//                 <td>${r.pending_qty}</td>
//                 <td>${r.available_qty}</td>
//                 <td>${r.ratio}%</td>
//                 <td>${r.delivery_date}</td>
//             </tr>`;
//         });

//         html += `
//         <tr style="font-weight:bold;background:#f3f3f3">
//             <td>Grand Total</td>
//             <td>${totals.ordered}</td>
//             <td>${totals.delivered}</td>
//             <td>${totals.pending}</td>
//             <td colspan="3"></td>
//         </tr>
//         </table>

//         <br>
//         <button class="btn btn-primary" onclick="print_pending_popup()">🖨 Print / Save PDF</button>

//         </div>`;

//         frappe.msgprint({
//             title: "Item Level Details",
//             message: html,
//             wide: true
//         });
//     }
// };


// window.print_pending_popup = function () {

//     let content = document.getElementById("pending-popup").innerHTML;

//     let w = window.open("", "", "width=900,height=700");

//     w.document.write(`
//         <html>
//         <head>
//             <title>Sales Order Fill Ratio</title>
//             <style>
//                 body { font-family: Arial; padding: 20px; }
//                 table { border-collapse: collapse; width: 100%; }
//                 table, th, td { border: 1px solid black; padding: 6px; text-align:center; }
//                 h3,h4,h5 { margin: 5px 0; }
//             </style>
//         </head>
//         <body>
//             ${content}
//         </body>
//         </html>
//     `);

//     w.document.close();
//     w.focus();
//     w.print();
// };

frappe.query_reports["SALES ORDER FILL RATIO"] = {

    filters: [
        { fieldname: "company", label: __("Company"), fieldtype: "Link", options: "Company", default: frappe.defaults.get_user_default("Company"), reqd: 1 },
        { fieldname: "from_date", label: __("From Date"), fieldtype: "Date" },
        { fieldname: "to_date", label: __("To Date"), fieldtype: "Date" },
        { fieldname: "warehouse", label: __("Warehouse"), fieldtype: "Link", options: "Warehouse" },
        { fieldname: "group_by_so", label: __("Group by Sales Order"), fieldtype: "Check", default: 1 }
    ],

    formatter(value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        if (column.fieldname === "pending_delivery" && data.pending_popup) {

            let safe_json = encodeURIComponent(data.pending_popup);

            return `<a style="font-weight:bold;color:#1674E0;cursor:pointer"
                onclick='frappe.query_reports["SALES ORDER FILL RATIO"]
                .show_popup("${safe_json}", "${data.customer}", "${data.sales_order}")'>
                View Items
            </a>`;
        }

        return value;
    },

    show_popup(encoded_data, customer, sales_order) {

        let data = JSON.parse(decodeURIComponent(encoded_data));
        let rows = data.items || [];
        let totals = data.totals || {};

        let html = `
        <div id="pending-popup">

        <h3>Sales Order Fill Ratio</h3>
        <h4>Customer: ${customer}</h4>
        <h5>Sales Order: ${sales_order}</h5>

        <div style="max-height:500px;overflow:auto">
        <table class="table table-bordered">
        <thead style="position:sticky;top:0;background:white">
        <tr>
            <th>Item</th>
            <th>Ordered</th>
            <th>Delivered</th>
            <th>Pending</th>
            <th>Available</th>
            <th>Fill %</th>
            <th>Date</th>
        </tr>
        </thead>
        <tbody>`;

        rows.forEach(r => {
            html += `
            <tr>
                <td>${r.item_name}</td>
                <td>${r.ordered_qty}</td>
                <td>${r.delivered_qty}</td>
                <td>${r.pending_qty}</td>
                <td>${r.available_qty}</td>
                <td>${r.ratio}%</td>
                <td>${r.delivery_date}</td>
            </tr>`;
        });

        html += `
        <tr style="font-weight:bold;background:#f3f3f3">
            <td>Grand Total</td>
            <td>${totals.ordered}</td>
            <td>${totals.delivered}</td>
            <td>${totals.pending}</td>
            <td colspan="3"></td>
        </tr>
        </tbody>
        </table>
        </div>

        <br>
        <button class="btn btn-primary" onclick="print_pending_popup()">🖨 Print / Save PDF</button>

        </div>`;

        let d = new frappe.ui.Dialog({
            title: "Item Level Details",
            fields: [{ fieldtype: "HTML", fieldname: "html" }]
        });

        d.fields_dict.html.$wrapper.html(html);
        d.show();

        // 🔥 make dialog wide
        d.$wrapper.find('.modal-dialog').css({
            "max-width": "95%",
            "width": "95%"
        });
    }
};

window.print_pending_popup = function () {

    let content = document.getElementById("pending-popup").innerHTML;

    let w = window.open("", "", "width=1000,height=800");

    w.document.write(`
        <html>
        <head>
            <title>Sales Order Fill Ratio</title>
            <style>
                body { font-family: Arial; padding: 20px; }
                table { border-collapse: collapse; width: 100%; }
                table, th, td { border: 1px solid black; padding: 6px; text-align:center; }
                h3,h4,h5 { margin: 5px 0; }
            </style>
        </head>
        <body>${content}</body>
        </html>
    `);

    w.document.close();
    w.focus();
    w.print();
};
